bin/i18n/utils/crowdin_client.rb

```diff
@@ -0,0 +1,221 @@
+require 'crowdin-api'
+
+require_relative '../i18n_script_utils'
+
+module I18n
+  module Utils
+    class CrowdinClient
+      RequestError = Class.new(StandardError)
+
+      CREDENTIALS_PATH = CDO.dir('bin/i18n/crowdin_credentials.yml').freeze
+      PROJECT_IDS = CDO.crowdin_project_ids.freeze
+      MAX_ITEMS_COUNT = Crowdin::Web::FetchAllExtensions::MAX_ITEMS_COUNT_PER_REQUEST.freeze
+      REQUEST_RETRY_ATTEMPTS = 2 # Number of retries for a failed request
+      REQUEST_RETRY_DELAY = 2 # Number of seconds to retry a failed request
+      RETRIABLE_ERRORS = [
+        '408 Request Timeout',
+        '429 Too Many Requests',
+        '500 Internal Server Error',
+        '503 Service Unavailable',
+      ].freeze # Request errors to retry
+
+      def initialize(project)
+        raise ArgumentError, 'project is invalid' unless PROJECT_IDS[project]
+
+        @project = project
+
+        @client = ::Crowdin::Client.new do |config|
+          config.api_token = YAML.load_file(CREDENTIALS_PATH)['api_token']
+          config.project_id = PROJECT_IDS[project]
+        end
+      end
+
+      # Creates the Crowdin directory for the given i18n source directory path
+      # https://developer.crowdin.com/api/v2/#operation/api.projects.directories.post
+      #
+      # @param dir_path [String] the i18n source directory path
+      # return [Hash, nil] the created Crowdin source directory data
+      def add_source_directory(dir_path)
+        crowdin_dir_path = crowdin_source_path(dir_path)
+        return if crowdin_dir_path.empty?
+
+        crowdin_dir_name = crowdin_source_name(crowdin_dir_path)
+        return if crowdin_dir_name.empty?
+
+        crowdin_parent_dir_path = File.dirname(crowdin_dir_path)
+        crowdin_parent_dir_id = source_directory(crowdin_parent_dir_path)&.dig('id')
+
+        request(:add_directory, directoryId: crowdin_parent_dir_id, name: crowdin_dir_name)['data']
+      end
+
+      # Retrieves the Crowdin source directory data of the given i18n source directory path
+      # https://developer.crowdin.com/api/v2/#operation/api.projects.directories.getMany
+      #
+      # @param dir_path [String] the i18n source directory path
+      # @return [Hash, nil] the Crowdin source directory data
+      def get_source_directory(dir_path)
+        crowdin_dir_path = crowdin_source_path(dir_path)
+        return if crowdin_dir_path.empty?
+
+        crowdin_dir_name = crowdin_source_name(crowdin_dir_path)
+        return if crowdin_dir_name.empty?
+
+        request_offset = 0
+        crowdin_directory = nil
+
+        loop do
+          crowdin_dirs = request(:list_directories, filter: crowdin_dir_name, offset: request_offset, limit: MAX_ITEMS_COUNT)['data']
+
+          crowdin_directory = crowdin_dirs.find {|crowdin_dir| crowdin_dir.dig('data', 'path') == crowdin_dir_path}
+
+          crowdin_directory || crowdin_dirs.size < MAX_ITEMS_COUNT ? break : request_offset += MAX_ITEMS_COUNT
```

Comment:
nit: this line is very clever, but I worry about its readability and long-term maintainability. Could you break this up (both here and in `get_source_file` below) and make it a bit more verbose and easier for someone new to this code to parse?